enterprise/cmd/frontend/internal/auth/oauth/middleware.go

```diff
@@ -66,7 +69,31 @@ func newOAuthFlowHandler(serviceType string) http.Handler {
 			http.Error(w, "Misconfigured GitHub auth provider.", http.StatusInternalServerError)
 			return
 		}
-		p.Login.ServeHTTP(w, req)
+
+		var extraScopes []string
+
+		// on Sourcegraph.com and for GitHub or Gitlab, check if the user has the AllowUserExternalServicePrivate tag
+		// and if so, ask the code host for additional scopes
+		if envvar.SourcegraphDotComMode() && (serviceType == "github" || serviceType == "gitlab") {
+			if actor := actor.FromContext(req.Context()); actor.IsAuthenticated() {
+				ok, err := database.Users(db).HasTag(req.Context(), actor.UID, database.TagAllowUserExternalServicePrivate)
+				if err != nil {
+					log15.Error("error while checking user tag", "id", actor.UID, "tag", database.TagAllowUserExternalServicePrivate)
+					http.Error(w, "Authentication failed. Try signing in again (and clearing cookies for the current site).", http.StatusInternalServerError)
+					return
+				}
+
+				if ok {
+					if serviceType == "github" {
```

Comment:
```suggestion
					if serviceType == extsvc.TypeGitHub {
```

Also suggest using `switch serviceType`, otherwise we will get "surprise" when GitHub and GitLab are not the only two supported code hosts.

Reply:
Good idea